core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -34,21 +46,182 @@ pub enum DiscoveryInitError {
 
 /// Responsible for communication with markets on other nodes
 /// during discovery phase.
-#[async_trait]
-pub trait Discovery: Send + Sync {
-    async fn bind_gsb(&self, prefix: String) -> Result<(), DiscoveryInitError>;
+#[derive(Clone)]
+pub struct Discovery {
+    inner: Arc<DiscoveryImpl>,
+}
+
+pub struct DiscoveryImpl {
+    offer_received: HandlerSlot<OfferReceived>,
```

Comment:
Consider wrapping here with `Arc`s
```suggestion
    offer_received: Arc<HandlerSlot<OfferReceived>>,
```

Reply:
Dependent on other discussions